kafka-connect-aws-s3/src/main/scala/io/lenses/streamreactor/connect/aws/s3/storage/AwsS3DirectoryLister.scala

```diff
@@ -0,0 +1,140 @@
+/*
+ * Copyright 2017-2023 Lenses.io Ltd
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.lenses.streamreactor.connect.aws.s3.storage
+
+import cats.effect.IO
+import cats.implicits._
+import io.lenses.streamreactor.connect.aws.s3.config.ConnectorTaskId
+import io.lenses.streamreactor.connect.aws.s3.model.location.RemoteS3RootLocation
+import software.amazon.awssdk.services.s3.S3Client
+import software.amazon.awssdk.services.s3.model._
+import software.amazon.awssdk.services.s3.paginators.ListObjectsV2Iterable
+
+import java.time.Clock
+import scala.collection.mutable
+import scala.jdk.CollectionConverters.IteratorHasAsScala
+import scala.jdk.CollectionConverters.ListHasAsScala
+
+abstract class AwsS3DirectoryLister(
+  implicit
+  connectorTask: ConnectorTaskId,
+  s3Client:      S3Client,
+) extends StorageInterface {
+  def findDirectories(
+    bucketAndPrefix:  RemoteS3RootLocation,
+    completionConfig: DirectoryFindCompletionConfig,
+    exclude:          Set[String],
+    continueFrom:     Option[String],
+  )(
+    implicit
+    clock: Clock,
+  ): IO[DirectoryFindResults] =
+    for {
+      listObjsReq <- createListObjectsRequest(bucketAndPrefix, continueFrom)
+      paginator   <- IO(s3Client.listObjectsV2Paginator(listObjsReq))
+      prefixInfo  <- extractPrefixesFromResponse(paginator, exclude, completionConfig)
+      flattened   <- flattenPrefixes(bucketAndPrefix, prefixInfo.partitions, completionConfig, exclude)
+    } yield {
+      prefixInfo match {
+        case CompletedDirectoryFindResults(_)        => CompletedDirectoryFindResults(flattened)
+        case a @ PausedDirectoryFindResults(_, _, _) => a.copy(flattened)
+      }
+    }
+
+  private def flattenPrefixes(
+    bucketAndPrefix:  RemoteS3RootLocation,
+    prefixes:         Set[String],
+    completionConfig: DirectoryFindCompletionConfig,
+    exclude:          Set[String],
+  )(
+    implicit
+    clock: Clock,
+  ): IO[Set[String]] =
+    if (completionConfig.levelsToRecurse == 0) IO.pure(prefixes)
```

Comment:
I would use `.delay` instead of `.pure` in general. The main issue with pure is that the passed param is eagerly evaluated, so any error produced there would raise the exception out of the IO/F context. 
We can use the regular apply for IO wich already delagates the construction to the delay method.
This is especially interesting/important when dealing with a Java api, since exceptions are part of their nature :)